### ISSUE TITLE
continuous deploy server so anyone can join

### ISSUE THREAD

**Issue by andrewgazelka** [hyperion-mc/hyperion#707]

**Comment by TestingPlant** [comment]
I could host the server for this. However my server runs aarch64 and newer rust versions seems to be getting an internal compile error, so I'll need to try to get that fixed first

**Comment by andrewgazelka** [comment]
> I could host the server for this. However my server runs aarch64 and newer rust versions seems to be getting an internal compile error, so I'll need to try to get that fixed first

oh cool what processor does it use?

**Comment by Kumpelinus** [comment]
I already offered my help as well (a 64 bit VPS). But afaik Andrew already has a VM for it.

**Comment by andrewgazelka** [comment]
> I could host the server for this. However my server runs aarch64 and newer rust versions seems to be getting an internal compile error, so I'll need to try to get that fixed first

oh yea look like flecs will not compile

```
 > [tag build-debug 1/1] RUN --mount=type=cache,target=/usr/local/cargo/registry     --mount=type=cache,target=/usr/local/cargo/git     --mount=type=cache,target=/app/target     cargo build --frozen &&     mkdir -p /app/build &&     cp target/debug/hyperion-proxy /app/build/ &&     cp target/debug/tag /app/build/:
0.528    Compiling idna_adapter v1.2.0
0.529    Compiling valence_generated v0.2.0-alpha.1+mc.1.20.1 (https://github.com/andrewgazelka/valence?branch=feat-open#7ed3252c)
0.531    Compiling flecs_ecs_sys v0.1.2 (https://github.com/Indra-db/Flecs-Rust#24b5a0de)
0.531    Compiling wayland-client v0.31.7
0.533    Compiling hyper v1.5.1
0.535    Compiling zbus_xml v4.0.0
0.537    Compiling hyperion-proto v0.1.0 (/app/crates/hyperion-proto)
0.539    Compiling wayland-csd-frame v0.3.0
0.539    Compiling zbus v4.4.0
0.541    Compiling wgpu-core v22.1.0
0.542    Compiling arboard v3.4.1
0.544    Compiling egui_glow v0.29.1
0.545    Compiling egui_dock v0.14.0
0.810    Compiling idna v1.0.3
1.123    Compiling zbus-lockstep v0.4.4
1.357    Compiling url v2.5.4
1.511    Compiling hyperion-proxy v0.1.0 (/app/crates/hyperion-proxy)
1.568    Compiling zbus-lockstep-macros v0.4.4
2.068    Compiling hyper-util v0.1.10
2.147    Compiling wayland-protocols v0.32.5
2.155    Compiling calloop-wayland-source v0.3.0
2.244    Compiling wayland-cursor v0.31.7
2.251    Compiling webbrowser v1.0.3
2.876    Compiling hyper-tls v0.6.0
2.994    Compiling hyper-rustls v0.27.3
3.057 error: failed to run custom build command for `flecs_ecs_sys v0.1.2 (https://github.com/Indra-db/Flecs-Rust#24b5a0de)`
3.058
3.058 Caused by:
3.058   process didn't exit successfully: `/app/target/debug/build/flecs_ecs_sys-411416b375e54bbc/build-script-build` (exit status: 101)
3.058   --- stdout
3.058   cargo:rerun-if-changed=src/flecs.h
3.058   cargo:rerun-if-changed=src/flecs.c
3.058   cargo:rerun-if-changed=src/flecs_rust.h
3.058   cargo:rerun-if-changed=src/flecs_rust.c
3.058   cargo:rerun-if-changed=build.rs
3.058   OUT_DIR = Some(/app/target/debug/build/flecs_ecs_sys-faa3143c600dd121/out)
3.058   OPT_LEVEL = Some(0)
3.058   TARGET = Some(aarch64-unknown-linux-gnu)
3.058   HOST = Some(aarch64-unknown-linux-gnu)
3.058   cargo:rerun-if-env-changed=CC_aarch64-unknown-linux-gnu
3.058   CC_aarch64-unknown-linux-gnu = None
3.058   cargo:rerun-if-env-changed=CC_aarch64_unknown_linux_gnu
3.058   CC_aarch64_unknown_linux_gnu = None
3.058   cargo:rerun-if-env-changed=HOST_CC
3.058   HOST_CC = None
3.058   cargo:rerun-if-env-changed=CC
3.058   CC = None
3.058   cargo:rerun-if-env-changed=CC_ENABLE_DEBUG_OUTPUT
3.058   RUSTC_WRAPPER = None
3.058   cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
3.058   CRATE_CC_NO_DEFAULTS = None
3.058   DEBUG = Some(true)
3.058   CARGO_CFG_TARGET_FEATURE = Some(aes,crc,lse,neon,sha2)
3.058   cargo:rerun-if-env-changed=CFLAGS_aarch64-unknown-linux-gnu
3.058   CFLAGS_aarch64-unknown-linux-gnu = None
3.058   cargo:rerun-if-env-changed=CFLAGS_aarch64_unknown_linux_gnu
3.058   CFLAGS_aarch64_unknown_linux_gnu = None
3.058   cargo:rerun-if-env-changed=HOST_CFLAGS
3.058   HOST_CFLAGS = None
3.058   cargo:rerun-if-env-changed=CFLAGS
3.058   CFLAGS = None
3.058   CARGO_ENCODED_RUSTFLAGS = Some(-Ctarget-cpu=native-Zshare-generics=y-Zthreads=8--cfgtokio_unstable)
3.058   cargo:rerun-if-env-changed=AR_aarch64-unknown-linux-gnu
3.058   AR_aarch64-unknown-linux-gnu = None
3.058   cargo:rerun-if-env-changed=AR_aarch64_unknown_linux_gnu
3.058   AR_aarch64_unknown_linux_gnu = None
3.058   cargo:rerun-if-env-changed=HOST_AR
3.058   HOST_AR = None
3.058   cargo:rerun-if-env-changed=AR
3.058   AR = None
3.058   cargo:rerun-if-env-changed=ARFLAGS_aarch64-unknown-linux-gnu
3.058   ARFLAGS_aarch64-unknown-linux-gnu = None
3.058   cargo:rerun-if-env-changed=ARFLAGS_aarch64_unknown_linux_gnu
3.058   ARFLAGS_aarch64_unknown_linux_gnu = None
3.058   cargo:rerun-if-env-changed=HOST_ARFLAGS
3.058   HOST_ARFLAGS = None
3.058   cargo:rerun-if-env-changed=ARFLAGS
3.058   ARFLAGS = None
3.058   cargo:rustc-link-lib=static=flecs
3.058   cargo:rustc-link-search=native=/app/target/debug/build/flecs_ecs_sys-faa3143c600dd121/out
3.058
3.058   --- stderr
3.058   thread 'main' panicked at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/bindgen-0.70.1/lib.rs:622:27:
3.058   Unable to find libclang: "couldn't find any valid shared libraries matching: ['libclang.so', 'libclang-*.so', 'libclang.so.*', 'libclang-*.so.*'], set the `LIBCLANG_PATH` environment variable to a path where one of these files can be found (invalid: [])"
3.058   note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
3.058 warning: build failed, waiting for other jobs to finish...
------
failed to sol
```

**Comment by TestingPlant** [comment]
Downgrading the toolchain with this commit works for me: https://github.com/TestingPlant/hyperion/commit/c201863a586ecb7632378b745989f466a95524ac

**Comment by TestingPlant** [comment]
> oh cool which processor does it use?

I'm using an Oracle VPS with 4 cores and 24 GB memory. It's using the VM.Standard.A1.Flex shape which uses Ampere Altra

**Comment by andrewgazelka** [comment]
oh nvm I got it to compile for aarch64 🤔

**Comment by andrewgazelka** [comment]
what error do you get when you do `docker compose up --build` @TestingPlant

**Comment by TestingPlant** [comment]
Tested it again and it somehow started working. Not sure why it works though.

Couldn't get docker working but this is what happened when I ran `cargo run --release --bin tag` previously
```
thread 'rustc' panicked at compiler/rustc_errors/src/lib.rs:646:17:               
`trimmed_def_paths` called, diagnostics were expected but none were emitted. Use `with_no_trimmed_paths` for debugging. Backtraces are currently disabled: set `RUST_BACKTRACE=1` and re-run to see where it happened.                                
stack backtrace:                      
   0:     0xf029740d01e4 - std::backtrace_rs::backtrace::libunwind::trace::h637ea50d17e411d7
                               at /rustc/d4025ee454169fbd22f5773f54348310ab6a47bb/library/std/src/../../backtrace/src/backtrace/libunwind.rs:116:5                  
   1:     0xf029740d01e4 - std::backtrace_rs::backtrace::trace_unsynchronized::h0b5fe6c9d9436388
                               at /rustc/d4025ee454169fbd22f5773f54348310ab6a47bb/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5                         
   2:     0xf029740d01e4 - std::sys::backtrace::_print_fmt::hd9738944a096f9d0     
                               at /rustc/d4025ee454169fbd22f5773f54348310ab6a47bb/library/std/src/sys/backtrace.rs:66:9    
   3:     0xf029740d01e4 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::hdd094dd7f6196825                              
                               at /rustc/d4025ee454169fbd22f5773f54348310ab6a47bb/library/std/src/sys/backtrace.rs:39:26   
   4:     0xf0297411cff8 - core::fmt::rt::Argument::fmt::ha726ffe17ac1731d        
                               at /rustc/d4025ee454169fbd22f5773f54348310ab6a47bb/library/core/src/fmt/rt.rs:177:76        
   5:     0xf0297411cff8 - core::fmt::write::haf914df1d5f049d1                    
                               at /rustc/d4025ee454169fbd22f5773f54348310ab6a47bb/library/core/src/fmt/mod.rs:1437:21      
   6:     0xf029740c4400 - std::io::Write::write_fmt::h40852d1a2a31069b           
                               at /rustc/d4025ee454169fbd22f5773f54348310ab6a47bb/library/std/src/io/mod.rs:1887:15        
   7:     0xf029740d0098 - std::sys::backtrace::BacktraceLock::print::h84d93da4875cc52d
                               at /rustc/d4025ee454169fbd22f5773f54348310ab6a47bb/library/std/src/sys/backtrace.rs:42:9    
   8:     0xf029740d23dc - std::panicking::default_hook::{{closure}}::heed26d4d69c909f9
                               at /rustc/d4025ee454169fbd22f5773f54348310ab6a47bb/library/std/src/panicking.rs:284:22      
   9:     0xf029740d2224 - std::panicking::default_hook::h4fe5cd6e53c7b923        
                               at /rustc/d4025ee454169fbd22f5773f54348310ab6a47bb/library/std/src/panicking.rs:311:9       
  10:     0xf0296e052ca0 - <alloc[1d4604b5e5222a9d]::boxed::Box<rustc_driver_impl[9d241e92d415777c]::install_ice_hook::{closure#0}> as core[494cfb034ae72dd8]::ops::function::Fn<(&dyn for<'a, 'b> core[494cfb034ae72dd8]::ops::function::Fn<(&'a std[4a649dfd30469b85]::panic::PanicHookInfo<'b>,), Output = ()> + core[494cfb034ae72dd8]::marker::Send + core[494cfb034ae72dd8]::marker::Sync, &std[4a649dfd30469b85]::panic::PanicHookInfo)>>::call             
  11:     0xf029740d2c1c - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h7b90e9ca40b0ce62              
                               at /rustc/d4025ee454169fbd22f5773f54348310ab6a47bb/library/alloc/src/boxed.rs:1984:9        
  12:     0xf029740d2c1c - std::panicking::rust_panic_with_hook::hd28c70d6f5623314
                               at /rustc/d4025ee454169fbd22f5773f54348310ab6a47bb/library/std/src/panicking.rs:825:13      
  13:     0xf029740d2908 - std::panicking::begin_panic_handler::{{closure}}::h48dcf4791a99ec7d
                               at /rustc/d4025ee454169fbd22f5773f54348310ab6a47bb/library/std/src/panicking.rs:690:13      
  14:     0xf029740d06dc - std::sys::backtrace::__rust_end_short_backtrace::hc0777be93a3d1ed6
                               at /rustc/d4025ee454169fbd22f5773f54348310ab6a47bb/library/std/src/sys/backtrace.rs:168:18  
  15:     0xf029740d25c8 - rust_begin_unwind                                      
                               at /rustc/d4025ee454169fbd22f5773f54348310ab6a47bb/library/std/src/panicking.rs:681:5       
  16:     0xf0296df56bf4 - core::panicking::panic_fmt::h8c9e76db71f91547          
                               at /rustc/d4025ee454169fbd22f5773f54348310ab6a47bb/library/core/src/panicking.rs:75:14      
  17:     0xf02973d31f2c - <rustc_errors[8cfb5e38a9127ac]::DiagCtxtInner as core[494cfb034ae72dd8]::ops::drop::Drop>::drop 
  18:     0xf0296e087f2c - core[494cfb034ae72dd8]::ptr::drop_in_place::<rustc_errors[8cfb5e38a9127ac]::DiagCtxt>           
  19:     0xf0296e089c88 - core[494cfb034ae72dd8]::ptr::drop_in_place::<rustc_session[a7744c7b40bb1b8a]::parse::ParseSess> 
  20:     0xf0296e08b0d4 - core[494cfb034ae72dd8]::ptr::drop_in_place::<rustc_interface[d89b66c0c02067aa]::interface::Compiler>                                     
  21:     0xf0296e084260 - <scoped_tls[7b797e7271203c50]::ScopedKey<rustc_span[f407efee92b1aa6b]::SessionGlobals>>::set::<rustc_interface[d89b66c0c02067aa]::util::run_in_thread_with_globals<rustc_interface[d89b66c0c02067aa]::util::run_in_thread_pool_with_globals<rustc_interface[d89b66c0c02067aa]::interface::run_compiler<(), rustc_driver_impl[9d241e92d415777c]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}, ()>                       
  22:     0xf0296e03f418 - rustc_span[f407efee92b1aa6b]::create_session_globals_then::<(), rustc_interface[d89b66c0c02067aa]::util::run_in_thread_with_globals<rustc_interface[d89b66c0c02067aa]::util::run_in_thread_pool_with_globals<rustc_interface[d89b66c0c02067aa]::interface::run_compiler<(), rustc_driver_impl[9d241e92d415777c]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}>                 
  23:     0xf0296e0544b8 - std[4a649dfd30469b85]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[d89b66c0c02067aa]::util::run_in_thread_with_globals<rustc_interface[d89b66c0c02067aa]::util::run_in_thread_pool_with_globals<rustc_interface[d89b66c0c02067aa]::interface::run_compiler<(), rustc_driver_impl[9d241e92d415777c]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>                    
  24:     0xf0296e03a8bc - <<std[4a649dfd30469b85]::thread::Builder>::spawn_unchecked_<rustc_interface[d89b66c0c02067aa]::util::run_in_thread_with_globals<rustc_interface[d89b66c0c02067aa]::util::run_in_thread_pool_with_globals<rustc_interface[d89b66c0c02067aa]::interface::run_compiler<(), rustc_driver_impl[9d241e92d415777c]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[494cfb034ae72dd8]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  25:     0xf029740dc47c - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h25b03a48e6eca4da     
                               at /rustc/d4025ee454169fbd22f5773f54348310ab6a47bb/library/alloc/src/boxed.rs:1970:9        
  26:     0xf029740dc47c - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h3d2e6fba6c19d449     
                               at /rustc/d4025ee454169fbd22f5773f54348310ab6a47bb/library/alloc/src/boxed.rs:1970:9        
  27:     0xf029740dc47c - std::sys::pal::unix::thread::Thread::new::thread_start::hb4618e76288d8c0f
                               at /rustc/d4025ee454169fbd22f5773f54348310ab6a47bb/library/std/src/sys/pal/unix/thread.rs:105:17                                     
  28:     0xf029770cd5c8 - start_thread  
                               at ./nptl/pthread_create.c:442:8                   
  29:     0xf02977135edc - thread_start  
                               at ./misc/../sysdeps/unix/sysv/linux/aarch64/clone.S:79
  30:                0x0 - <unknown>     

error: the compiler unexpectedly panicked. this is a bug.                         

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md                            

note: please make sure that you have updated to the latest nightly                

note: please attach the file at `/home/remote-dev/.cargo/git/checkouts/flecs-rust-ff75a97755951fdf/24b5a0d/flecs_ecs/rustc-ice-2024-12-13T20_15_28-730047.txt` to your bug report       

note: compiler flags: --crate-type lib -C opt-level=3 -C embed-bitcode=no -C strip=debuginfo -C target-cpu=native          

note: some of the compiler flags provided by cargo are hidden                     

query stack during panic:             
end of query stack                    
error: could not compile `flecs_ecs` (lib)                                        

Caused by:                            
  process didn't exit successfully: `/home/remote-dev/.rustup/toolchains/nightly-aarch64-unknown-linux-gnu/bin/rustc --crate-name flecs_ecs --edition=2021 /home/remote-dev/.cargo/git/checkouts/flecs-rust-ff75a97755951fdf/24b5a0d/flecs_ecs/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=153 --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C embed-bitcode=no --warn=unused_lifetimes '--warn=clippy::semicolon_if_nothing_returned' '--warn=clippy::print_stdout' '--warn=clippy::print_stderr' '--warn=clippy::float_cmp_const' '--warn=clippy::float_cmp' '--warn=clippy::doc_markdown' --cfg 'feature="default"' --cfg 'feature="flecs_alerts"' --cfg 'feature="flecs_app"' --cfg 'feature="flecs_doc"' --cfg 'feature="flecs_http"' --cfg 'feature="flecs_json"' --cfg 'feature="flecs_log"' --cfg 'feature="flecs_manual_registration"' --cfg 'feature="flecs_meta"' --cfg 'feature="flecs_metrics"' --cfg 'feature="flecs_module"' --cfg 'feature="flecs_os_api_impl"' --cfg 'feature="flecs_pipeline"' --cfg 'feature="flecs_regenerate_binding_c"' --cfg 'feature="flecs_rest"' --cfg 'feature="flecs_script"' --cfg 'feature="flecs_snapshot"' --cfg 'feature="flecs_stats"' --cfg 'feature="flecs_system"' --cfg 'feature="flecs_timer"' --cfg 'feature="flecs_units"' --check-cfg 'cfg(docsrs)' --check-cfg 'cfg(feature, values("default", "flecs_alerts", "flecs_app", "flecs_disable_build_c", "flecs_doc", "flecs_force_build_debug_c", "flecs_force_build_release_c", "flecs_force_enable_ecs_asserts", "flecs_http", "flecs_journal", "flecs_json", "flecs_log", "flecs_manual_registration", "flecs_meta", "flecs_metrics", "flecs_module", "flecs_nightly_tests", "flecs_os_api_impl", "flecs_perf_trace", "flecs_pipeline", "flecs_regenerate_binding_c", "flecs_rest", "flecs_script", "flecs_snapshot", "flecs_stats", "flecs_system", "flecs_term_count_64", "flecs_timer", "flecs_units", "flecs_use_os_alloc", "test-with-crash-handler"))' -C metadata=636094ee05582202 -C extra-filename=-e352faeb2e98d34e --out-dir /home/remote-dev/hyperion/target/release/deps -C strip=debuginfo -L dependency=/home/remote-dev/hyperion/target/release/deps --extern bitflags=/home/remote-dev/hyperion/target/release/deps/libbitflags-9dc40629b0c994f5.rmeta --extern compact_str=/home/remote-dev/hyperion/target/release/deps/libcompact_str-c73f1c99b71f9342.rmeta --extern flecs_ecs_derive=/home/remote-dev/hyperion/target/release/deps/libflecs_ecs_derive-b3dc8f9d81f1e47e.so --extern flecs_ecs_sys=/home/remote-dev/hyperion/target/release/deps/libflecs_ecs_sys-94ebb75b6b61ab83.rmeta --extern hashbrown=/home/remote-dev/hyperion/target/release/deps/libhashbrown-29c77b652dca04fa.rmeta --cap-lints allow --cfg tokio_unstable -Ctarget-cpu=native -L native=/home/remote-dev/hyperion/target/release/build/flecs_ecs_sys-e69f4edddbd8e9e2/out` (exit status: 101)
warning: build failed, waiting for other jobs to finish...                        
```
Not really sure why it works now but I'm not that concerned about it

**Comment by andrewgazelka** [comment]
ahhhh yea I had that too. i had to downgrade the rust-toolchain because of that as well

also added packages now... but only for `amd64` for now
<img width="276" alt="image" src="https://github.com/user-attachments/assets/5c7261a1-ff58-4bb3-84e5-61e3fd5e3ac4" />

**Comment by andrewgazelka** [comment]
want to look into how we would continuously deploy with GitHub actions → your server perhaps?

**Comment by andrewgazelka** [comment]
right now we are using https://github.com/andrewgazelka/hyperion/blob/89c2ef439b2e1df6310a6c37e81a4441005a99d1/.github/workflows/publish.yml#L1-L53

**Comment by TestingPlant** [comment]
> want to look into how we would continuously deploy with GitHub actions → your server perhaps? we need to get `arm64` servers hopefully https://depot.dev/ will pull thru and sponsor this project so we can use their service for free 🤞 cause building arm64 thru emulator on GitHub will suckkkkk

We could set up a repository webhook for the push event which gets my server to compile the new code. I'm getting the VPS through always free resources so I'm not worried about any compute costs

**Comment by andrewgazelka** [comment]
also I think https://github.com/andrewgazelka/hyperion/pull/711

**Comment by andrewgazelka** [comment]
@TestingPlant lmk when you get this working

**Comment by TestingPlant** [comment]
Implemented in #733
